blog/content/post/2024-11-26-measured-boot-improvements.md

```diff
@@ -0,0 +1,222 @@
+the event log to the OS reside. It is done in [this Pull
+Request](https://github.com/Dasharo/edk2/pull/139).
+
+Of course, this is not the end of modifications. Depending on the payload,
+some glue may also be required on the coreboot side. We have to teach coreboot
+not to publish TPM ACPI tables (TCPA and TPM2) because they contain the
+address of the TPM event log in the memory. This task is now left to the
+payload, which publishes the respective TPM ACPI table with its own, combined
+TPM event log. It is achieved in the [previously linked Pull Request on EDKII
+side](https://github.com/Dasharo/edk2/pull/139) and also on [coreboot
+side](https://github.com/Dasharo/coreboot/pull/517) (some of the changes here
+are already upstream). Besides the TPM ACPI table, coreboot creates an event
+log using only one hashing algorithm:
+
+- SHA1 - default for TPM1.2
+- SHA256 - default for TPM2.0
+
+TPM2.0 log format supports agile measurements, i.e. there may be multiple
+measurements of the same component using multiple hashing algorithms. For
+simplicity, coreboot uses only one currently. The choice of the hashing
+algorithm must also be passed to the UEFI Payload. It is done by [specifying a
+mask of supported hashing algorithms at UEFI Payload build
+time](https://github.com/Dasharo/coreboot/pull/517/files#diff-58bb01776a00af5a5d61de00d6ea3008bcd67177fb7991fb41e8976c6b4d046a).
+
+Now that all the pieces are in place, we can enjoy having a consistent and
+complete TPM event log in the Dasharo, effectively solving the [issue of not
+reproducible PCR2
+value](https://github.com/Dasharo/dasharo-issues/issues/455).
+
+## Summary
+
+TPM event log integrity is an excellent achievement for our Dasharo team. While we
+did an enormous work to improve it, there are still some aspects not covered
+by current implementation:
+
+1. What if the PCR banks are changed?
+
+   The EDKII TPM2 setup menu allows changing PCR banks, effectively
+   influencing what measurements (precisely hash algorithms used to measure
+   firmware components) may be recorded in the TPM. coreboot is hardcoded to
+   use only one hashing algorithm (fixed at build). It may result in PCR banks
+   not being properly extended or the PCR values being entirely out of sync
+   with the TPM event log. coreboot should query the TPM for currently enabled
+   PCR banks and use all algorithms enabled at the time. That, of course, may
+   have various implications:
+
+    - Pre-RAM measurements are saved in Cache-as-RAM (CAR) memory, which is
+      limited. Calculating more hashes will cause the CAR memory to shrink,
+      which effectively limits the amount of cache for silicon initialization
+      code. It may not be a big deal for modern platforms with big caches, but
+      rather for the old ones.
+    - coreboot supports only one algorithm in a single event log entry.
+      Support for an agile format would have to be implemented.
```

Comment:
What would be the current mitigation for that?

Can't we just set up an admin password so a change of PCR banks would not be possible for someone who doesn't know the password?

Reply:
In general even if an admin changes PCR banks, the change won't be applied or it will break measured boot.

The mitigation is like I described: be agile and calculate hashes for all currently enabled banks. Of course coreboot changes to event logging must follow.

Reply:
> In general even if an admin changes PCR banks, the change won't be applied or it will break measured boot.

So why are we not saying what mitigates that? Dear admin, do not change banks because it will break the system's trustworthiness. The admin must comply with that policy and should be informed about it. If the admin doesn't want users to change that policy, even accidentally, then prevent that by protecting the option with a password. That is the current mitigation. There should be a better one with the correct implementation, but that still needs to be implemented. We may be working on that. Or could you have a reference point to follow up?